teps/0119-simplifying-tektons-compute-resources-api.md

```diff
@@ -0,0 +1,119 @@
+---
+status: implementable
+title: Simplifying Tekton's Compute Resources API
+creation-date: '2022-08-16'
+last-updated: '2022-08-16'
+authors:
+- '@lbernick'
+---
+
+# TEP-0119: Simplifying Tekton's Compute Resources API
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+    - [Sidecars](#sidecars)
+    - [Naming](#naming)
+<!-- /toc -->
+
+## Summary
+
+This TEP simplifies the API fields available for configuring compute resource requirements of a TaskRun.
+
+## Motivation
+
+Originally, users had only the option of specifying compute resources for each Step and Sidecar in a Task.
+There are two issues with this approach:
+1. Compute resources are a runtime requirement, not an authoring time requirement.
+[TEP-0094](./0094-configuring-resources-at-runtime.md) added the ability to specify compute resources for each Step and Sidecar
+at runtime (i.e. on a TaskRun).
+2. Since Steps run sequentially, but Kubernetes assumes that all containers run in parallel, specifying compute resources
+for each Step rather than for the TaskRun as a whole leads to confusing behavior during scheduling.
+[TEP-0104](./0104-tasklevel-resource-requirements.md) addressed this problem by adding the ability to specify the total
+compute resources for all the Steps in a TaskRun.
+
+The goal of this TEP is to simplify Tekton's compute resource API to remove redundancy and lessen confusion.
+
+### Goals
+
+- Remove redundant or confusing API fields related to compute resources.
+- Consistent naming for all API fields related to compute resources.
+
+### Non-Goals
+
+- Discussion of how Tekton handles [LimitRanges](https://kubernetes.io/docs/concepts/policy/limit-range/)
+or [ResourceQuotas](https://kubernetes.io/docs/concepts/policy/resource-quotas/).
+
+## Proposal
+
+- Remove `task.spec.steps[].resources` (beta). This field should be marked as deprecated in v1beta1 and not be ported to v1.
+- Promote `taskRun.spec.computeResources` to beta, and make it available in v1 behind a "beta" feature flag.
```

Comment:
Hmmm, does this mean that users of the v1 api version will have no way of specifying computeResources when they have set api features to "stable" ?

Reply:
Yes, but this is already the case since we are not stabilizing task.spec.steps.resources (https://github.com/tektoncd/community/pull/769)

Reply:
From a user perspective though, this feels odd  - I had some way of specifying compute resources via step resources (despite all the caveats) in v1beta1 when I had stable set, but if I upgrade to v1 I can no longer do so without also having to opt in to all the other beta features

Reply:
I fully agree-- but I think the alternatives are less good. I've added this as a drawback and added some notes on what I see as our alternatives. I'm wondering if this is more a result of how we've chosen to do api versioning?

Reply:
Agree. I think it's more to do with how we do our feature stability across alpha, beta, and stable.